File: src/oned/rss/FinderPattern.java
/*
 * Copyright 2009 ZXing authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.zxing.oned.rss;

import com.google.zxing.RXingResultPoint;

/**
 * Encapsulates an RSS barcode finder pattern, including its start/end position and row.
 */
public final class FinderPattern {

  private final int value;
  private final int[] startEnd;
  private final RXingResultPoint[] resultPoints;

  public FinderPattern(int value, int[] startEnd, int start, int end, int rowNumber) {
    this.value = value;
    this.startEnd = startEnd;
    this.resultPoints = new RXingResultPoint[] {
        new RXingResultPoint(start, rowNumber),
        new RXingResultPoint(end, rowNumber),
    };
  }

  public int getValue() {
    return value;
  }

  public int[] getStartEnd() {
    return startEnd;
  }

  public RXingResultPoint[] getRXingResultPoints() {
    return resultPoints;
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof FinderPattern)) {
      return false;
    }
    FinderPattern that = (FinderPattern) o;
    return value == that.value;
  }

  @Override
  public int hashCode() {
    return value;
  }

}

File: src/oned/rss/finder_pattern.rs
/*
 * Copyright 2009 ZXing authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::hash::Hash;

use crate::{point, Point};

/**
 * Encapsulates an RSS barcode finder pattern, including its start/end position and row.
 */
#[derive(Clone)]
pub struct FinderPattern {
    value: u32,
    startEnd: [usize; 2],
    resultPoints: Vec<Point>,
}

impl FinderPattern {
    pub fn new(value: u32, startEnd: [usize; 2], start: usize, end: usize, rowNumber: u32) -> Self {
        Self {
            value,
            startEnd,
            resultPoints: vec![
                point(start as f32, rowNumber as f32),
                point(end as f32, rowNumber as f32),
            ],
        }
    }

    pub const fn getValue(&self) -> u32 {
        self.value
    }

    pub fn getStartEnd(&self) -> &[usize] {
        &self.startEnd
    }

    #[cfg(all(test, feature = "image"))]
    pub(crate) fn getStartEndMut(&mut self) -> &mut [usize] {
        &mut self.startEnd
    }

    pub fn getPoints(&self) -> &[Point] {
        &self.resultPoints
    }
}

impl PartialEq for FinderPattern {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}
impl Eq for FinderPattern {}
impl Hash for FinderPattern {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}
